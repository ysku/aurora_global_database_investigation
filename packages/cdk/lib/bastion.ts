import {
  GenericLinuxImage,
  Instance,
  InstanceClass,
  InstanceSize,
  InstanceType,
  SecurityGroup,
  Vpc,
} from '@aws-cdk/aws-ec2';
import { ManagedPolicy, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { App, Stack, StackProps } from '@aws-cdk/core';

export type BastionProps = {
  dbAccessSG: SecurityGroup;
  vpc: Vpc;
} & StackProps;

export class BastionStack extends Stack {
  constructor(scope: App, id: string, props: BastionProps) {
    super(scope, id, props);

    const { dbAccessSG, vpc } = props;

    new Instance(this, 'bastion', {
      instanceName: 'bastion',
      vpc,
      instanceType: InstanceType.of(InstanceClass.T2, InstanceSize.MICRO),
      machineImage: new GenericLinuxImage({
        'ap-northeast-1': 'ami-059b6d3840b03d6dd', // Ubuntu Server 20.04 LTS (HVM), SSD Volume Type
      }),
      securityGroup: dbAccessSG,
      role: new Role(this, 'bastion-role', {
        assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
        managedPolicies: [
          ManagedPolicy.fromAwsManagedPolicyName(
            'AmazonSSMManagedInstanceCore'
          ),
        ],
      }),
    });
  }
}
