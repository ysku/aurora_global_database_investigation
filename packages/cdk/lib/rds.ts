import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  SecurityGroup,
  SubnetType,
  Vpc,
} from '@aws-cdk/aws-ec2';
import {
  AuroraPostgresEngineVersion,
  Credentials,
  DatabaseCluster,
  DatabaseClusterEngine,
  ParameterGroup,
} from '@aws-cdk/aws-rds';
import { Secret } from '@aws-cdk/aws-secretsmanager';
import { App, Stack, StackProps } from '@aws-cdk/core';

export type RDSStackProps = {
  dbClusterIdentifier: string;
  dbName: string;
  dbSG: SecurityGroup;
  dbUserName: string;
  vpc: Vpc;
} & StackProps;

export class RDSStack extends Stack {
  readonly dbCluster: DatabaseCluster;
  readonly secret: Secret;

  constructor(scope: App, id: string, props: RDSStackProps) {
    super(scope, id, props);

    const { dbClusterIdentifier, dbName, dbSG, dbUserName, vpc } = props;

    // use secret manager to configure database username and password
    // cf. https://docs.aws.amazon.com/ja_jp/secretsmanager/latest/userguide/intro.html
    this.secret = new Secret(this, 'Secret', {
      secretName: `${id}Secret`,
      description: 'RDS database auto-generated user password',
      generateSecretString: {
        excludeCharacters: ':/?#[]@!$&"\'()*+,;=%{}|\\^~[]`',
        generateStringKey: 'password',
        passwordLength: 32,
        secretStringTemplate: `{"username": "${dbUserName}"}`,
      },
    });

    const parameterGroup = new ParameterGroup(this, 'RDSParameterGroup', {
      description: `parameter group for ${dbName}`,
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_10_7,
      }),
      parameters: {
        log_statement: 'all',
      },
    });

    this.dbCluster = new DatabaseCluster(this, 'RDS', {
      clusterIdentifier: dbClusterIdentifier,
      // cannot use upper case characters.
      defaultDatabaseName: dbName,
      parameterGroup: parameterGroup,
      instances: 2,
      instanceProps: {
        instanceType: InstanceType.of(InstanceClass.R4, InstanceSize.LARGE),
        vpc: vpc,
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED,
        },
        securityGroups: [dbSG],
      },
      port: 5432,
      engine: DatabaseClusterEngine.auroraPostgres({
        version: AuroraPostgresEngineVersion.VER_10_7,
      }),
      credentials: Credentials.fromSecret(this.secret),
    });
  }
}
