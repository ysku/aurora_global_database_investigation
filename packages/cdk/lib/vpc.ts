import {
  CfnSecurityGroupIngress,
  SecurityGroup,
  SubnetType,
  Vpc,
  VpcProps,
} from '@aws-cdk/aws-ec2';
import { App, Stack, StackProps } from '@aws-cdk/core';

export class VPCStack extends Stack {
  readonly vpc: Vpc;
  readonly dbAccessSG: SecurityGroup;
  readonly dbSG: SecurityGroup;
  readonly instanceSG: SecurityGroup;

  constructor(scope: App, id: string, props?: StackProps) {
    super(scope, id, props);

    const vpcProps: VpcProps = {
      cidr: '192.168.0.0/16',
      enableDnsHostnames: true,
      enableDnsSupport: true,
      maxAzs: 3,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'public',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'private',
          subnetType: SubnetType.PRIVATE,
        },
        {
          cidrMask: 24,
          name: 'isolated',
          subnetType: SubnetType.ISOLATED,
        },
      ],
    };

    // NOTE: default security group will be also created
    this.vpc = new Vpc(this, 'VPC', vpcProps);

    this.dbAccessSG = new SecurityGroup(this, 'dbaccesssg', {
      vpc: this.vpc,
      description: 'for instances that access database',
      securityGroupName: 'Database Access',
    });

    this.dbSG = new SecurityGroup(this, 'dbsg', {
      vpc: this.vpc,
      description: 'for database',
      securityGroupName: 'Database',
    });

    this.instanceSG = new SecurityGroup(this, 'instancesg', {
      vpc: this.vpc,
      description: 'for ec2 instance',
      securityGroupName: 'Instance',
    });

    // allow DatabaseSG to access from DBAccessSG
    new CfnSecurityGroupIngress(this, 'fromdbaccesssg', {
      groupId: this.dbSG.securityGroupId,
      ipProtocol: 'tcp',
      description: `from instances with sg named ${this.dbAccessSG.securityGroupName}`,
      fromPort: 5432,
      toPort: 5432,
      sourceSecurityGroupId: this.dbAccessSG.securityGroupId,
    });

    // allow DatabaseSG to access from DBAccessSG
    new CfnSecurityGroupIngress(this, 'frominstance', {
      groupId: this.dbSG.securityGroupId,
      ipProtocol: 'tcp',
      description: `from instances with sg named ${this.instanceSG.securityGroupName}`,
      fromPort: 5432,
      toPort: 5432,
      sourceSecurityGroupId: this.instanceSG.securityGroupId,
    });
  }
}
