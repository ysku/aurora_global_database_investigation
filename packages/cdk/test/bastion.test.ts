import { SynthUtils } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';

import { BastionStack } from '../lib/bastion';
import { VPCStack } from '../lib/vpc';

describe('BastionStack', () => {
  const app = new App();
  const vpc = new VPCStack(app, 'TestVPCStack');

  test('snapshot', () => {
    const stack = new BastionStack(app, 'TestBastionStack', {
      dbAccessSG: vpc.dbAccessSG,
      vpc: vpc.vpc,
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
