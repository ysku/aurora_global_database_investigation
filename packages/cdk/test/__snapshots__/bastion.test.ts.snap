// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BastionStack snapshot 1`] = `
Object {
  "Mappings": Object {
    "bastionAmiMapE6BABB0F": Object {
      "ap-northeast-1": Object {
        "ami": "ami-059b6d3840b03d6dd",
      },
    },
  },
  "Resources": Object {
    "bastion34F27305": Object {
      "DependsOn": Array [
        "bastionroleFD8E05AF",
      ],
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "IamInstanceProfile": Object {
          "Ref": "bastionInstanceProfileBF7A08EB",
        },
        "ImageId": Object {
          "Fn::FindInMap": Array [
            "bastionAmiMapE6BABB0F",
            Object {
              "Ref": "AWS::Region",
            },
            "ami",
          ],
        },
        "InstanceType": "t2.micro",
        "SecurityGroupIds": Array [
          Object {
            "Fn::ImportValue": "TestVPCStack:ExportsOutputFnGetAttdbaccesssg59FD6FC1GroupIdA22CF994",
          },
        ],
        "SubnetId": Object {
          "Fn::ImportValue": "TestVPCStack:ExportsOutputRefVPCprivateSubnet1SubnetE522290D7D3DE097",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "bastion",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "bastionInstanceProfileBF7A08EB": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "bastionroleFD8E05AF",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "bastionroleFD8E05AF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
